File: DesignPatterns/Structural/Facade/Java/HotelClient.java
interface Menu {
  public String foodItems();
}

class VegMenu implements Menu {

  public String foodItems() {
    return "This is a VegMenu";
  }
}

class NonVegMenu implements Menu {

  public String foodItems(){
    return "This is a NonVegMenu";
  }
}

class VegNonVegMenu implements Menu {

  public String foodItems() {
    return "This is a VegNonVegMenu";
  }
}

interface Hotel {
  public Menu getMenus();
}

class VegRestaurant implements Hotel {

  public Menu getMenus() {
    Menu m = new VegMenu();
    return m;
  }
}

class NonVegRestaurant implements Hotel {
  
  public Menu getMenus() {
    Menu m = new NonVegMenu();
    return m;
  }
}

class VegNonVegRestaurant implements Hotel {

  public Menu getMenus() {
    Menu m = new VegNonVegMenu();
    return m;
  }
}

class HotelKeeper {

  public Menu getVegMenu() {
    Hotel h = new VegRestaurant();
    Menu m =  h.getMenus();
    return m; 
  }

  public Menu getNonVegMenu() {
    Hotel h = new NonVegRestaurant();
    Menu m = h.getMenus();
    return m;
  }

  public Menu getVegNonVegMenu() {
    Hotel h = new VegNonVegRestaurant();;
    Menu m = h.getMenus();
    return m;
  }
}


public class HotelClient {
  public static void main(String[] args) {
    HotelKeeper hk = new HotelKeeper();
    System.out.println(hk.getVegMenu().foodItems());
    System.out.println(hk.getNonVegMenu().foodItems());
    System.out.println(hk.getVegNonVegMenu().foodItems());
  }
}

File: DesignPatterns/Structural/Facade/Ruby/hotel_client.rb
class Menu
  def food_items
    raise NotImplementedError
  end
end

class VegMenu < Menu
  def food_items
    'This is a VegMenu'
  end
end

class NonVegMenu < Menu
  def food_items
    'This is a NonVegMenu'
  end
end

class VegNonVegMenu < Menu
  def food_items
    'This is a VegNonVegMenu'
  end
end

class Restaurant
  def fetch_menu
    raise NotImplementedError
  end
end

class VegRestaurant < Restaurant
  def fetch_menu
    VegMenu.new
  end
end

class NonVegRestaurant < Restaurant
  def fetch_menu
    NonVegMenu.new
  end
end

class VegNonVegRestaurant < Restaurant
  def fetch_menu
    VegNonVegMenu.new
  end
end


class HotelKeeper
  def fetch_veg_menu
    VegRestaurant.new.fetch_menu
  end

  def fetch_nonveg_menu
    NonVegRestaurant.new.fetch_menu
  end

  def fetch_vegnonveg_menu
    VegNonVegRestaurant.new.fetch_menu
  end
end


hk = HotelKeeper.new
puts hk.fetch_veg_menu.food_items
puts hk.fetch_nonveg_menu.food_items
puts hk.fetch_vegnonveg_menu.food_items